### ISSUE TITLE
Provide generic -DTARGET_* defines

### ISSUE THREAD

**Issue by nalajcie** [phoenix-rtos/phoenix-rtos-build#60]
As of now, most repos do it in target-specific Makefile's and they sometimes differentiate between the repos.

I'm talking about:
```make
CFLAGS += -DTARGET_IMXRT

ifneq (, $(findstring 117, $(TARGET)))
  CFLAGS += -DTARGET_IMXRT1170
else ifneq (, $(findstring 105, $(TARGET)))
  CFLAGS += -DTARGET_IMXRT1050
else ifneq (, $(findstring 106, $(TARGET)))
  CFLAGS += -DTARGET_IMXRT1060
endif
```
which seems to be repeated in most repos.

They should either be moved intact to generic makefiles or we should introduce alternative which would cover all future bases, eg:
```make
CFLAGS += -DTARGET_FAMILY=$(TARGET_FAMILY) -DTARGET_SUBFAMILY=$(TARGET_SUBFAMILY)
```

@agkaminski, @Maxez, @anglov I would like to know Your input on that

**Comment by anglov** [comment]
AFAIK it's mainly used by kernel and devices. Application could use it in hacky way to get board type (as in the most case every board have other CPU :)) ) but it use TARGET_FAMILY and TARGET_SUBFAMILY

IDK if this variables is really used outside of kernel/devices/usb?

**Comment by nalajcie** [comment]
> AFAIK it's mainly used by kernel and devices. Application could use it in hacky way to get board type (as in the most case every board have other CPU :)) ) but it use TARGET_FAMILY and TARGET_SUBFAMILY
> 
> IDK if this variables is really used outside of kernel/devices/usb?

use cases outside of kernel/devices:
- libgraph: https://github.com/phoenix-rtos/phoenix-rtos-corelibs/pull/3/files#diff-0c697798e39d131cf7f6d7827f522d05a22cf194207a1532ef71d46bf40ab112R20
- defined but not used currently in: filesystems, usb
- `TARGET_STM32*` is being set in several places, but I don't see it being used anywhere?

**Comment by anglov** [comment]
Yeah, I think that TARGET_STM32* was used when multidrv was same for l1 and l4 (before splitting).

Ok, libgraph is good example when it would be needed in application code.

For me I resolve everything in Makefiles (other files for host other for real target) and by third TARGET variable

**Comment by Maxez** [comment]
I used TARGET_RISCV64 in quite a few places for virtio library and drivers development (due to some problems with user interrupt handlers on riscv64). It was more convenient to use #ifdef TARGET_RISCV64 instead of hiding platform specific implementation in separate module and resolving compilation at makefile level.

Right now, after many lectures from Paweł on #ifdef's being bad, I learned not to rely on them. I found __attribute__((weak)) functions a nice way of dealing with the problem (I used it in libgraph - it no longer relies on TARGET_IA32 macro).

To sum things up, we shouldn't encourage #ifdef TARGET_X usage + there are other ways to write platform specific code so I'm against these defines. Maybe we could even remove them altogether (from kernel and devices)?

**Comment by agkaminski** [comment]
@nalajcie I guess fixed in https://github.com/phoenix-rtos/phoenix-rtos-build/commit/bee018405d14154c938704c105a702f6683e7795?

**Comment by nalajcie** [comment]
Yeah, closing.
